rust/src/countme/cookie.rs

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use anyhow::{anyhow, bail, Result};
+use chrono::prelude::*;
+use openat_ext::OpenatDirExt;
+use serde::{Deserialize, Serialize};
+use std::io::Read;
+
+/// State directory used to store the countme cookie
+const STATE_DIR: &str = "/var/lib/rpm-ostree-countme";
+/// Cookie file name
+const COUNTME_COOKIE: &str = "countme";
+
+/// Width of the sliding time window (in seconds): 1 week
+const COUNTME_WINDOW: i64 = 7 * 24 * 60 * 60;
+/// Starting point of the sliding time window relative to the UNIX epoch:
+/// Monday (1970-01-05 00:00:00 UTC)
+/// Allows for aligning the window with a specific weekday
+const COUNTME_OFFSET: i64 = 345600;
+
+/// Cookie v0 JSON file format
+///
+/// The values stored in this cookie are the same as the ones used in the DNF
+/// Count Me Cookie with the exception of the COUNTME_BUDGET which is not
+/// included as this implementation directly queries all repo metalinks once:
+///   - epoch: the timestamp of the first ever counted window
+///   - window: the timestamp of the last successfully counted window
+///
+/// See the libdnf implementation at:
+/// https://github.com/rpm-software-management/libdnf/blob/95b88b141a3f97feb94eadb2480f6857b6d1fcae/libdnf/repo/Repo.cpp#L1038
+///
+/// This cookie is marked as version 0 similarly to how libdnf defines a version
+/// for their cookie to simplify potential future updates but no additional code
+/// has been added yet to handle multiple versions as we do not need it yet.
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+struct CookieV0 {
+    epoch: i64,
+    window: i64,
+}
+
+impl CookieV0 {
+    /// Try to parse a string as a v0 cookie
+    fn new(string: &str) -> Result<Self> {
+        let c: CookieV0 = serde_json::from_str(string)?;
+        if c.epoch <= 0 {
+            bail!("Invalid value for epoch: '{}'", c.epoch);
+        }
+        if c.window <= 0 {
+            bail!("Invalid value for window: '{}'", c.window);
+        }
+        Ok(c)
+    }
+
+    /// Export a v0 cookie to JSON for persistent storage
+    fn to_json(epoch: i64, window: i64) -> Result<String> {
+        match serde_json::to_string(&CookieV0 { epoch, window }) {
+            Err(e) => Err(anyhow!(e)),
```

Comment:
Doesn't it work to just use `serde_json::to_string(&CookieV0 { epoch, window })?` here?  The `?` should do the conversion for many error types.

Reply:
I tried:
```
error[E0308]: mismatched types
  --> rust/src/countme/cookie.rs:72:9
   |
71 |     fn to_json(epoch: i64, window: i64) -> Result<String> {
   |                                            -------------- expected `std::result::Result<std::string::String, anyhow::Error>` because of return type
72 |         serde_json::to_string(&CookieV0 { epoch, window })
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `anyhow::Error`, found struct `serde_json::Error`
   |
   = note: expected enum `std::result::Result<_, anyhow::Error>`
              found enum `std::result::Result<_, serde_json::Error>`
```
and:
```
error[E0308]: try expression alternatives have incompatible types
  --> rust/src/countme/cookie.rs:72:9
   |
72 |         serde_json::to_string(&CookieV0 { epoch, window })?
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         expected enum `std::result::Result`, found struct `std::string::String`
   |         help: try using a variant of the expected enum: `lockfile::_::_serde::export::Ok(serde_json::to_string(&CookieV0 { epoch, window })?)`
   |
   = note: expected enum `std::result::Result<std::string::String, anyhow::Error>`
            found struct `std::string::String`
```
without success so I went with this verbose option.

Reply:
Try `Ok(serde_json::to_string(&CookieV0 { epoch, window })?)`.

Reply:
Feels so obvious right now! 😅

Reply:
Oh no worries, I still get confused around Rust error type conversions.